sklearn/cluster/_k_means_fast.pyx

```diff
@@ -7,155 +7,256 @@
 #         Lars Buitinck
 #
 # License: BSD 3 clause
-#
-# cython: boundscheck=False, wraparound=False, cdivision=True
 
-from libc.math cimport sqrt
 import numpy as np
-import scipy.sparse as sp
 cimport numpy as np
 cimport cython
 from cython cimport floating
+from libc.math cimport sqrt
+
+from ..utils.extmath import row_norms
+
+
+np.import_array()
 
-from ..utils.sparsefuncs_fast import assign_rows_csr
-from ..utils._cython_blas cimport _dot
 
 ctypedef np.float64_t DOUBLE
 ctypedef np.int32_t INT
 
 
-np.import_array()
+cdef floating _euclidean_dense_dense(floating* a,
+                                     floating* b,
+                                     int n_features,
+                                     bint squared) nogil:
+    """Euclidean distance between a dense and b dense"""
+    cdef:
```

Comment:
Maybe add a comment on when one would use this as opposed to a^2 + b^2 - 2 a.dot(b) with a BLAS call, as was done before in this file.

Reply:
I'd say never. The 1D version of the dot trick is not faster than the exact computation. I made this helper for everywhere I needed euclidean distances between 1D arrays. I don't like putting a comment saying that there is an alternative but I don't advise to use it.